DM-46799: refactor DatasetRecordStorageManager

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

TallJimbo committed:
Fix variable-binding bug in GeneralQueryResults.

TallJimbo committed:
Minor doc fixes.

TallJimbo committed:
Remove vestiges of autoincrement integer dataset ID support.

The lone concrete dataset storage manager class and its intermediate
base class have been merged, and many function parameters (all
internal, of course) that plumb the SQL dataset ID type through many
layers have been dropped.

TallJimbo committed:
Move methods from DatasetRecordStorage to its manager.

This is a mostly-mechanical change; for each method:

- move a method to the manager class;

- give it a dataset type argument, and have it look up the storage
  object internally;

- change calling code (SqlRegistry) to stop looking up the storage
  object before calling it.

The ultimate goal is to get rid of DimensionRecordStorageManager, at
least as a public interface.  But the steps after this one won't be
nearly as mechanical, so it's useful to separate them.

TallJimbo committed:
Clarify save_dimension_graph behavior in docs.

TallJimbo committed:
Don't pass dataset types when only their dimensions are needed.

TallJimbo committed:
Rework dataset manager and dataset type caching.

The dataset type cache now holds just the dataset type definition and
dataset ID by name, with the SQLAlchemy table objects instead cached
by DimensionGroup so they can be shared by multiple dataset types.

DatasetRecordStorage has been removed (both the base class and its
sole implementation) - its methods had already been moved to
DatasetRecordStorageManager, and it no longer works as the opaque
thing to put in the cache.  Instead there's a new subpackage-private
DynamicTables class that is cached by DimensionGroup (this is where
the lazy loading of SQLAlchemy table objects now happens), and a
module-private _DatasetRecordStorage struct that just combines that
with the dataset type and its ID, to make it more convenient to pass
all of that around.

I also threw in some changes to the insert/import implementations
because I started trying to reduce the degree that
DatasetRecordStorage was being passed things that were either totally
unused or assumed (without checking) have some value.  I quickly
realized that this problem is ubiquitous (especially with storage
classes) and should be a separate ticket, but I've kept what I already
did since I think it's a step in the right direction.

TallJimbo committed:
Fix old DimensionGraph nomenclature reference.

TallJimbo committed:
Ban multiple-dataset-type results in general find-first queries.

Supporting these would be extra complexity I don't think we need.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_datasets.py:
```diff
@@ -701,3 +381,292 @@ def fetch_summaries(
+
+        Notes
+        -----
+        Associating a dataset with into collection that already contains a
```
```suggestion
        Associating a dataset into collection that already contains a
```

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_datasets.py:
```diff
@@ -701,3 +381,292 @@ def fetch_summaries(
+
+    @abstractmethod
+    def disassociate(
+        self, dataset_type: DatasetType, collection: CollectionRecord, datasets: Iterable[DatasetRef]
```
Do you have to specify `DataType` here, `DatasetRef` should already know its `DataType`?

TallJimbo replied:
This is one of the cases I just didn't want to add to the scope of the ticket, because while you're right, there was just so much of it that it would have been a distraction.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_datasets.py:
```diff
@@ -701,3 +381,292 @@ def fetch_summaries(
+            the same `DatasetType` and data ID and an overlapping validity
+            range.
+        DatasetTypeError
+            Raised if ``self.datasetType.isCalibration() is False``.
```
```suggestion
            Raised if ``dataset_type.isCalibration() is False``.
```

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_datasets.py:
```diff
@@ -701,3 +381,292 @@ def fetch_summaries(
+        Raises
+        ------
+        DatasetTypeError
+            Raised if ``self.datasetType.isCalibration() is False``.
```
```suggestion
            Raised if ``dataset_type.isCalibration() is False``.
```

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_datasets.py:
```diff
@@ -701,3 +381,292 @@ def fetch_summaries(
+        columns: Set[str],
+        context: SqlQueryContext,
+    ) -> Relation:
+        """Return a `sql.Relation` that represents a query for for this
```
```suggestion
        """Return a `sql.Relation` that represents a query for this
```

andy-slac commented on python/lsst/daf/butler/registry/_caching_context.py:
```diff
@@ -54,10 +54,16 @@ class is passed to the relevant managers that can use it to query or
+
+    `GenericCachingContext` is generic over two kinds of opaque dataset type
+    data, with the expectation that most code will use the ``CachingContext``
+    type alias (which resolves to `GenericCachingContext[object, object]``);
```
```suggestion
    type alias (which resolves to `GenericCachingContext[object, object]`);
```

andy-slac commented on python/lsst/daf/butler/registry/_dataset_type_cache.py:
```diff
@@ -33,68 +33,96 @@
-        self._cache = {}
+        self._by_name_cache = {}
+        self._by_dimensions_cache = {}
         self._full = False
```
Maybe also `self._dimensions_full = False`?

andy-slac commented on python/lsst/daf/butler/registry/datasets/byDimensions/_manager.py:
```diff
@@ -53,39 +55,52 @@
+    """Information cached about a dataset type.
+
+    This combines information cached with different keys - the dataset type
+    and its ID are cached by name, while the tables are cache by the dataset
```
```suggestion
    and its ID are cached by name, while the tables are cached by the dataset
```

andy-slac commented on python/lsst/daf/butler/registry/sql_registry.py:
```diff
@@ -1376,18 +1344,19 @@ def certify(self, collection: str, refs: Iterable[DatasetRef], timespan: Timespa
-            collection or if one or more datasets are of a dataset type for
-            which `DatasetType.isCalibration` returns `False`.
+        DatasetTypeError
+            Raised if ``self.datasetType.isCalibration() is False``.
```
```suggestion
            Raised if ``ref.datasetType.isCalibration() is False`` for any ref.
```

andy-slac commented on python/lsst/daf/butler/registry/sql_registry.py:
```diff
@@ -1425,24 +1394,25 @@ def decertify(
-            Raised if ``collection`` is not a `~CollectionType.CALIBRATION`
-            collection or if ``datasetType.isCalibration() is False``.
+        DatasetTypeError
+            Raised if ``self.datasetType.isCalibration() is False``.
```
```suggestion
            Raised if ``datasetType.isCalibration() is False``.
```

andy-slac reviewed (approved):
Looks great, few docstring fixes. Thanks for fixing GeneralQueryResult!

TallJimbo committed:
Switch back to returning given DatasetType in registry imports.

When the given dataset type differs from the registered dataset type
in imports, it's not clear what the ideal behavior is, but the right
choice for *this* ticket is clearly to not change that behavior.
